versioned/storage/cleanup/src/main/java/org/projectnessie/versioned/storage/cleanup/CutHistoryImpl.java

```diff
@@ -0,0 +1,127 @@
+import org.projectnessie.versioned.storage.common.persist.Persist;
+
+public class CutHistoryImpl implements CutHistory {
+  private final CutHistoryParams context;
+  private final RateLimit scanRateLimiter;
+
+  public CutHistoryImpl(CutHistoryParams context, IntFunction<RateLimit> rateLimitIntFunction) {
+    this.context = context;
+    this.scanRateLimiter = rateLimitIntFunction.apply(context.scanObjRatePerSecond());
+  }
+
+  @Override
+  public CutHistoryScanResult identifyAffectedCommits() {
+    Builder result = CutHistoryScanResult.builder();
+
+    Persist persist = context.persist();
+    ObjId cutPoint = context.cutPoint();
+
+    long numObjects = 0;
+    // Identify commits, whose "tails" overlap the cut point.
+    // scanAllCommitLogEntries() returns all commits in no specific order, parents may be scanned
+    // before or after their children.
+    try (CloseableIterator<Obj> scan = persist.scanAllObjects(Collections.singleton(COMMIT))) {
+      while (scan.hasNext()) {
+        scanRateLimiter.acquire();
+
+        CommitObj commit = (CommitObj) scan.next();
+        numObjects++;
+        if (commit.tail().contains(cutPoint)) {
+          result.addAffectedCommitId(commit.id());
+        }
+      }
+
+      result.numScannedObjs(numObjects);
+      return result.build();
+    }
+  }
+
+  @Override
+  public UpdateParentsResult rewriteParents(CutHistoryScanResult scanResult) {
+    ImmutableUpdateParentsResult.Builder result = UpdateParentsResult.builder();
+    CommitLogic commitLogic = commitLogic(context.persist());
+    for (ObjId id : scanResult.affectedCommitIds()) {
+      try {
+        CommitObj commitObj = commitLogic.fetchCommit(id);
+        if (commitObj == null) {
+          result.putFailure(id, new IllegalAccessException("Commit not found: " + id));
+          continue;
+        }
+
+        if (commitObj.tail().isEmpty()) {
+          continue;
+        }
+
+        CommitObj updated =
+            CommitObj.commitBuilder()
+                .from(commitObj)
+                .tail(commitObj.tail().subList(0, 1)) // one direct parent for simplicity
+                .build();
+
+        if (!context.dryRun()) {
+          context.persist().upsertObj(updated);
+        }
+
+      } catch (Exception e) {
+        result.putFailure(id, e);
+      }
+    }
+    return result.build();
+  }
+
+  @Override
+  public UpdateParentsResult cutHistory() {
+    ObjId id = context.cutPoint();
+    ImmutableUpdateParentsResult.Builder result = UpdateParentsResult.builder();
+    CommitLogic commitLogic = commitLogic(context.persist());
+    try {
+      CommitObj commitObj = commitLogic.fetchCommit(id);
+      if (commitObj == null) {
+        result.putFailure(id, new IllegalAccessException("Commit not found: " + id));
+      } else {
+        CommitObj updated =
+            CommitObj.commitBuilder()
+                .from(commitObj)
+                .secondaryParents(List.of())
+                .tail(List.of())
```

Comment:
IIRC there's never an empty list - EMPTY_OBJ_ID must be present

Reply:
added `EMPTY_OBJ_ID`